tests/system/Database/Builder/ReplaceTest.php

```diff
@@ -53,4 +54,36 @@ public function testReplaceThrowsExceptionWithNoData()
 
         $builder->replace();
     }
+
+    public function testReplaceBatch()
```

Comment:
I think a live test is also needed to test all supported databases.

Reply:
This would rule out using ON DUPLICATE KEY UPDATE with INSERT but it seems REPLACE is a MySQL extention.

This shouldn't affect the replace() method if it only works with MySQL to begin with.

https://dev.mysql.com/doc/refman/8.0/en/replace.html

PostgreSQL uses a similar ON CONFLICT DO UPDATE 
https://www.postgresqltutorial.com/postgresql-tutorial/postgresql-upsert/

I guess the proper term here would be UPSERT

Reply:
> I guess the proper term here would be UPSERT

Good point. REPLACE and UPSERT are different.

REPLACE = INSERT or (DELETE + INSERT)
UPSERT = INSERT or UPDATE

Reply:
> > I guess the proper term here would be UPSERT
> 
> Good point. REPLACE and UPSERT are different.
> 
> REPLACE = INSERT or (DELETE + INSERT) UPSERT = INSERT or UPDATE

Looking at the Postgre implementation of replace() they use a second query to see if the record exists and if it does then it updates it, otherwise inserts it. This would be an upsert. For that matter they should have used ON CONFLICT DO UPDATE.
https://www.prisma.io/dataguide/postgresql/inserting-and-modifying-data/insert-on-conflict

Personally I'd prefer the UPSERT operation to the REPLACE. The current implementation has muddied the difference.  

I currently have a need to upsert 1700 records at a time pretty regularly and was looking to give CI this functionality. I never use MySQL's REPLACE so I got looking into the difference and decided it might cause too many problems. 

I've used Soql's upsert function a lot as well. With Soql you run an primary or external key. The nice thing though with mysql is that it runs on any unamed constraint.

If we are going to get picky on UPSERT vs REPLACE than maybe I should pursue developing upsert and upsertBatch. I don't know that we need both though and having both might just bloat CI.

Also, could use some help with the other databases. Would it be improper to throw an error not yet implemented in some of them? Then as we get around to it we add that functionality. That or maybe an insert ignore query followed by an update query.

Reply:
> Personally I'd prefer the UPSERT operation to the REPLACE. The current implementation has muddied the difference.

I agree with it. 
Oracle implementation uses `MERGE` that is UPSERT.

> Also, could use some help with the other databases. Would it be improper to throw an error not yet implemented in some of them? Then as we get around to it we add that functionality.

Implementing a new feature for all database drivers by one person is too hard.
Personally I agree with using an error not yet implemented.
But the new feature can be implemented in all supported database, and needs to be approved.

Reply:
I've done some research on all the current databases and think I could put them all together. 

The Oracle driver is almost done already with the current replace method. It just needs to pass multiple values by select * from Dual Union All.

Postgre and SQLite both use a very similar ON CONFLICT DO.. 

The standard SQL specification:
"Database management systems Oracle Database, IBM Db2, Teradata, EXASOL, Firebird, CUBRID, H2, HSQLDB, MS SQL, Vectorwise and Apache Derby support the standard syntax."
https://en.wikipedia.org/wiki/Merge_(SQL)

So MSSQL should work the same of very similar to Oracle. The MERGE should be the BaseBuilder method. 

Mysql is mostly done already.

So question is, should I create new methods upsert() and upsertBatch() and leave alone the replace() implementation? If I were to change to upsert, should we scrap the PR and create a new one?